### DISCUSSION TITLE
In Windows, it's not possible to switch to kana input while an application is running.

### DISCUSSION THREAD

**Discussion by mikunimaru** [google/mozc#806] [Ideas]
## Description
In Windows, the switch between kana input and romaji input is only effective within applications launched after the switch. This behavior can be extremely inconvenient and a potential source of trouble for kana input users. On the contrary, it is also inconvenient for romaji input users who have mistakenly set it to kana input and need to revert the setting. In contrast, MSIME allows for switching between kana input and romaji input even while applications are running.

## Steps to reproduce
1. Change the input method to kana input.

## Expected behavior
Kana input will become available in the currently running app.

## Actual behavior
The currently running app will remain in the romaji input mode, and only the apps launched afterwards will allow kana input.

## Environment
 - OS: Windows

**Comment by yukawa** [comment]
> The currently running app will remain in the romaji input mode, and only the apps launched afterwards will allow kana input.

We are aware of this limitation and that's why show the following warning dialog.
<img width="415" alt="romaji-kana-settings" src="https://github.com/google/mozc/assets/5861397/12cb4876-a1b8-4d73-b77c-7148c083cb82">

I agree that supporting this scenario is technically possible, but doing so would most likely require some non-trivial amount of changes from the IPC protocol level. Also, as Mozc is a multi-platform IME, we want to make sure that the way to address your request is something that is feasible in macOS, `ibus_mozc`, and `emacs-mozc`.

For Windows, what we need to do would be updating the following logic.
https://github.com/google/mozc/blob/f79e092ac7068f6bbdacb2b069641771e7cb1f43/src/win32/base/keyevent_handler.cc#L477-L517
Currently `KeyEvent::key_string` is sent when and only when Kana input is selected, and the reason the new settings becomes available only for newly launched applications is that `behavior.prefer_kana_input` is loaded only when `mozc_tip64.dll` is loaded.

To make the setting dynamically updatable, I think we may want to always and unconditionally set `KeyEvent::key_string` in `mozc_tip64.dll` then let `mozc_server` decide how to use it.
https://github.com/google/mozc/blob/f79e092ac7068f6bbdacb2b069641771e7cb1f43/src/protocol/commands.proto#L260-L266

To summarize, action items to make this happen are:
1. Add a new parameter to `commands.proto` so that Kana handling can be fully done in the `mozc_server`
2. Update `mozc_server` to support the new parameter.
3. Update `mozc_tip32.dll` and `mozc_tip64.dll` to support the new parameter for Windows.
4. Update mac client, ibus client, and emacs client to support new parameter.
5. Update `mozc_tool` to remove the warning dialog.

I don't say we would never implement the above change, but I think it's not likely to happen in the near future.

**Comment by mikunimaru** [comment]
I found something interesting about this issue. When pressing Alt+Katakana/Hiragana key, it's possible to switch the input method while the application is running. In Windows, after changing the input method to "かな入力 (切り替え可)" or "ローマ字入力 (切り替え可)", it's possible to switch the input method while the application is running using the mentioned method. Regarding Linux, I remember being able to switch the input method on fcitx5-mozc by simply pressing Alt+Katakana/Hiragana keys during the application's launch without making any changes to the settings.
![image](https://hoshino-wp.com/wp-content/uploads/2019/09/123-768x363.jpg)

Would it be challenging to unify the processing for Alt+Hiragana/Katakana key and the processing when changing the input method from the settings screen?

**Reply by yukawa** [reply]
> When pressing Alt+Katakana/Hiragana key, it's possible to switch the input method while the application is running. In Windows, after changing the input method to "かな入力 (切り替え可)" or "ローマ字入力 (切り替え可)", it's possible to switch the input method while the application is running using the mentioned method.

No, the above scenario and immediately reflecting global settings are two different things.

Basically the challenge is how Mozc TIP DLLs (`mozc_tip32.dll`, `mozc_tip64.dll`), which run inside each application (e.g. notepad, Chrome, ...), can get notified when the settings are updated. Currently Mozc TIP DLLs fetch these settings only once when they are initialized to avoid unnecessary IPCs. So to achieve your request there are basically two approaches:

A. Let Mozc TIP DLLs fetch settings more frequently (e.g. every time you switch window focus)
B. Make Mozc TIP DLLs more stateless so that TIP DLLs can always behave in the same way regardless of the Romaji/Kana settings.

In our design philosophy, B is the way to go. It requires relatively complicated changes as I explained above though.

Doing A might be easy but it introduces more IPCs, which keeps affecting all the users even after the initial setup is done. That's why we didn't go with B, and just introduced a warning dialog instead.

> Regarding Linux, I remember being able to switch the input method on fcitx5-mozc by simply pressing Alt+Katakana/Hiragana keys during the application's launch without making any changes to the settings.

I haven't tried Alt+Katakana/Hiragan in fcitx5-mozc, but it sounds like that's an unexpected behavior. Before introducing  "かな入力 (切り替え可)" and "ローマ字入力 (切り替え可)", we had heard so many complaints from Google Japanese Input users who accidentally switched to Kana mode then had no idea how to fix it. IIUC, MS-IME in Windows 11 also disables Alt+Katakana/Hiragan by default.
<img width="339" alt="image" src="https://github.com/google/mozc/assets/5861397/5965d3fc-202a-4cff-bb57-3eecfef84edc">
